Método Simplex
===

**Juan David Velásquez Henao**  
jdvelasq@unal.edu.co   
Universidad Nacional de Colombia, Sede Medellín  
Facultad de Minas  
Medellín, Colombia

---

Haga click [aquí](https://github.com/jdvelasq/metodos-y-modelos/blob/master/02.04-metodo-simplex.ipynb) para acceder a la última versión online

Haga click [aquí](http://nbviewer.jupyter.org/github/jdvelasq/metodos-y-modelos/blob/master/02.04-metodo-simplex.ipynb) para ver la última versión online en `nbviewer`. 

---

### Sistema canónico

$$ 
\begin{array}{cccccccc}
  +x_4 &     &     & -3x_1 &  +x_2 &      & = & 3\\
       & x_5 &     &  -x_1 &  +x_2 & -x_3 & = & 0\\
       &     & x_6 &  +x_1 & +2x_2 &      & = & 6\\
\end{array}
$$

* $x_4$, $x_5$ y $x_6$ son llamadas variables básica o dependientes y están asociadas al vector $b$.
* $x_1$, $x_2$ y $x_3$	son las variables no básicas o independientes.
* El valor de las variables básicas se obtiene al asignar valores a las variables no básicas.
* Si las variables no básicas son cero y $b_i \ge 0$, para $i=1,...,m$, los valores de las variables básicas forman una solución básica factible.


### Método simplex

Función objetivo:

$$\text{max } x_1 + 3x_2$$

Sujeto a:
$$ \begin{array}{cccc}
   -x_1 & +x_2 & \le & 1\\
   +x_1 & +x_2 & \le & 2\\
    x_1,&  x_2 & \ge & 0
  \end{array}$$

**Forma estándar.** Se convierte en un problema de minimización: $\text{min } z = -x_1 - 3x_2$

$$ 
\begin{array}{ccccccc}
   -x_1 &  +x_2 & +x_3 &      &    & = & 1\\
   +x_1 &  +x_2 &      & +x_4 &    & = & 2\\
   -x_1 & -3x_2 &      &      & -z & = & 0
\end{array}
$$
  
SBF: $x_1 = x_2 = 0$, $x_3 = 1$, $x_4 = 2$, $z=0$  

Note que las variables básicas tienen coeficiente cero en la función objetivo.

In [46]:
import numpy as np
a = np.matrix('-1. 1 1 0 1; 1 1 0 1 2; -1 -3 0 0 0')
a

matrix([[-1.,  1.,  1.,  0.,  1.],
        [ 1.,  1.,  0.,  1.,  2.],
        [-1., -3.,  0.,  0.,  0.]])

**Paso 1.** Defina cuál variable ($x_1$, $x_2$) debe entrar a la base, tal que la solución siga siendo factible. Recuerde que $x_1$, $x_2$, $x_3$, $x_4 \ge 0$.


* Si $x_1$ se hace diferente de cero, entonces  decrece en: $z=0-x_1$.


* Si $x_2$ se hace diferente de cero, entonces $z$ decrece en: $z=0-3x_2$. 


* Conclusión: se consigue disminuir más $z$ si $x_2$ ingresa a la base.


* Nota 1. Si no hay coeficientes negativos la solución es óptima.


* Nota 2. Si la solución es óptima y una variable no básica tiene coeficiente cero, entonces existen múltiples soluciones.

**Paso 2.** Defina cuál variable ($x_3$ o $x_4$) debe salir de la base. ¿Qué tan grande puede hacerse $x_2$? Recuerde que $x_3$, $x_4 \ge = 0$.

* De la restricción 1: $x_3 = 1 - x_2$ ($x_2$ puede tomar un valor máximo de 1, ya que $x_3 \ge 0$)


* De la restricción 2: $x_4 = 2 - x_2$ ($x_2$ puede tomar un valor máximo de 2, ya que $x_4 \ge 0$)


* Conclusión: El máximo valor que puede tomar $x_2$ es 1 debido a la restrición 1. Esto es, $x_2$ reemplaza a $x_3$ en la base.

**Paso 3**. Mediante operaciones elementales se obtiene la nueva base.

In [47]:
# se hace E2 <- E2 - E1 y E3 <- E3 + 3*E1
a[1,:] = a[1,:] + (-1) * a[0,:]
a[2,:] = a[2,:] + 3 * a[0,:]
a

matrix([[-1.,  1.,  1.,  0.,  1.],
        [ 2.,  0., -1.,  1.,  1.],
        [-4.,  0.,  3.,  0.,  3.]])

$$ 
\begin{array}{ccccccc}
   -x_1 & +x_2 &  +x_3 &      &    & = & 1\\
  +2x_1 &      &  -x_3 & +x_4 &    & = & 1\\
  -4x_1 &      & +3x_3 &      & -z & = & 3
\end{array}
$$
  
SBF: $x_1 = x_3 = 0$, $x_2 = 1$, $x_4 = 1$, $z=-3$  

** Paso 4.** Retorne al Paso 1. 

---

**Continuación**. Se aplica nuevamente el Paso 1 y se encuentra que $x_1$ reemplaza a $x_4$ debido a la restricción 2.

In [48]:
a[1,:] = a[1,:] / 2          ## se hace E2 <- E2 / 2
a[0,:] = a[0,:] + a[1,:]     ## E1 <- E1 + E2
a[2,:] = a[2,:] + 4 * a[1,:] ## E3 <- E3 + 4*E2
a

matrix([[ 0. ,  1. ,  0.5,  0.5,  1.5],
        [ 1. ,  0. , -0.5,  0.5,  0.5],
        [ 0. ,  0. ,  1. ,  2. ,  5. ]])

$$ 
\begin{array}{ccccccc}
        & +x_2 &  +\frac{1}{2} x_3 & \frac{1}{2} x_4 &    & = & \frac{3}{2}\\
    x_1 &      &  -\frac{1}{2} x_3 & \frac{1}{2} x_4 &    & = & \frac{1}{2}\\
        &      &              +x_3 &           2 x_4 & -z & = & 5
\end{array}
$$
  
SBF: $x_1 = 0.5$, $x_2 = 1.5$, $x_3 = x_4 = 0$, $z=-5$  

Ya que todos los coeficientes de las variables no básicas ($x_3$ y $x_4$) son no negativos ($\ge 0$), el algoritmo convergio a una solución óptima. Y ya que las variables no básicas tiene coeficientes diferentes de cero, la solución es única.

**Problema 1.** Resuelva el siguiente problema usando el método simplex.

Función objetivo:

$$\text{max } 40x_1 + 60x_2$$

Sujeto a:
$$ \begin{array}{cccc}
    2x_1 &  +x_2 & \le & 70\\
     x_1 &  +x_2 & \le & 40\\
     x_1 & +3x_2 & \le & 90\\
        & x_1, x_2    &  \ge & 0
  \end{array}$$

**Problema 2.** Resuelva el siguiente problema usando el método simplex.

Función objetivo:

$$\text{max } 3x_1 + 2x_2$$

Sujeto a:
$$ 
\begin{array}{cccc}
    6x_1 &  +4x_2 & \le & 24\\
     x_1 &        & \le & 3\\
         & x_1, x_2 &  \ge & 0
\end{array}
$$

Método Simplex
===

**Juan David Velásquez Henao**  
jdvelasq@unal.edu.co   
Universidad Nacional de Colombia, Sede Medellín  
Facultad de Minas  
Medellín, Colombia

---

Haga click [aquí](https://github.com/jdvelasq/metodos-y-modelos/blob/master/02.04-metodo-simplex.ipynb) para acceder a la última versión online

Haga click [aquí](http://nbviewer.jupyter.org/github/jdvelasq/metodos-y-modelos/blob/master/02.04-metodo-simplex.ipynb) para ver la última versión online en `nbviewer`. 

---